File: content/snippets/go/interview_lcci/17/15_longest_word.go
package main

import "sort"

func longestWord(words []string) string {
	wordSet := make(map[string]bool)
	for _, w := range words {
		wordSet[w] = true
	}

	var canBuild func(w string) bool
	canBuild = func(w string) bool {
		if w == "" {
			return true
		}
		for i := 1; i <= len(w); i++ {
			if wordSet[w[:i]] && canBuild(w[i:]) {
				return true
			}
		}
		return false
	}

	sort.Slice(words, func(i, j int) bool {
		if len(words[i]) != len(words[j]) {
			return len(words[i]) > len(words[j])
		}
		return words[i] < words[j]
	})

	for _, w := range words {
		delete(wordSet, w)
		if canBuild(w) {
			return w
		}
		wordSet[w] = true
	}
	return ""
}

File: content/snippets/typescript/interview_lcci/17/15_longest_word.ts
function longestWord(words: string[]): string {
    const wordSet = new Set(words);

    function canBuild(w: string): boolean {
        if (w === '') return true;
        for (let i = 1; i <= w.length; i++) {
            if (wordSet.has(w.slice(0, i)) && canBuild(w.slice(i))) {
                return true;
            }
        }
        return false;
    }

    words.sort((a, b) => b.length - a.length || a.localeCompare(b));

    for (const w of words) {
        wordSet.delete(w);
        if (canBuild(w)) return w;
        wordSet.add(w);
    }
    return '';
}
